File: FlashlightsInTheDark_MacOS/Model/EventRecipe.swift
import Foundation

// MARK: - Event Recipe Bundle -------------------------------------------------
/// Top-level bundle describing all pre-authored event recipes.
public struct EventRecipeBundle: Decodable {
    public let source: String?
    public let eventCount: Int?
    public let generated: String?
    public let events: [EventRecipe]
}

// MARK: - Event Recipe --------------------------------------------------------
public struct EventRecipe: Identifiable, Decodable {
    public let id: Int
    /// Official trigger-score measure for this event, not the sung-note onset.
    public let measure: Int?
    /// Official trigger-score beat position for this event, not the sung-note onset.
    public let position: String?
    public let primerAssignments: [PrimerColor: PrimerAssignment]
    public let lighting: EventLighting?

    enum CodingKeys: String, CodingKey {
        case id, measure, position, primer, lighting
    }

    public init(
        id: Int,
        measure: Int?,
        position: String?,
        primerAssignments: [PrimerColor: PrimerAssignment],
        lighting: EventLighting? = nil
    ) {
        self.id = id
        self.measure = measure
        self.position = position
        self.primerAssignments = primerAssignments
        self.lighting = lighting
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(Int.self, forKey: .id)
        measure = try container.decodeIfPresent(Int.self, forKey: .measure)
        position = try container.decodeIfPresent(String.self, forKey: .position)
        if let rawPrimer = try container.decodeIfPresent([String: PrimerAssignment].self, forKey: .primer) {
            var mapped: [PrimerColor: PrimerAssignment] = [:]
            for (key, assignment) in rawPrimer {
                if let color = PrimerColor(rawValue: key.lowercased()) {
                    mapped[color] = assignment
                }
            }
            primerAssignments = mapped
        } else {
            primerAssignments = [:]
        }
        lighting = try container.decodeIfPresent(EventLighting.self, forKey: .lighting)
    }
}

// MARK: - Lighting ------------------------------------------------------------
public struct EventLighting: Decodable {
    public let summary: String?
    public let scoreDynamics: String?
    public let designTags: [String]
    public let durationMs: Double?
    public let parts: [LightStaff: EventLightPartPlan]

    enum CodingKeys: String, CodingKey {
        case summary
        case scoreDynamics
        case designTags
        case durationMs
        case parts
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        summary = try container.decodeIfPresent(String.self, forKey: .summary)
        scoreDynamics = try container.decodeIfPresent(String.self, forKey: .scoreDynamics)
        designTags = try container.decodeIfPresent([String].self, forKey: .designTags) ?? []
        durationMs = try container.decodeIfPresent(Double.self, forKey: .durationMs)

        if let rawParts = try container.decodeIfPresent([String: EventLightPartPlan].self, forKey: .parts) {
            var mapped: [LightStaff: EventLightPartPlan] = [:]
            for (key, value) in rawParts {
                if let staff = LightStaff(rawValue: key) {
                    mapped[staff] = value
                }
            }
            parts = mapped
        } else {
            parts = [:]
        }
    }
}

public struct EventLightPartPlan: Decodable {
    public let label: String
    public let summary: String
    public let motion: String
    public let peakLevel: Double
    public let durationMs: Double
    public let keyframes: [EventLightKeyframe]
}

public struct EventLightKeyframe: Decodable {
    public let atMs: Double
    public let level: Double
}

public enum LightStaff: String, CaseIterable, Codable, Identifiable {
    case sopranoL1 = "soprano_l1"
    case sopranoL2 = "soprano_l2"
    case tenorL = "tenor_l"
    case bassL = "bass_l"
    case altoL2 = "alto_l2"
    case altoL1 = "alto_l1"

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .sopranoL1: return "Sop-L1"
        case .sopranoL2: return "Sop-L2"
        case .tenorL: return "Ten-L"
        case .bassL: return "Bass-L"
        case .altoL2: return "Alto-L2"
        case .altoL1: return "Alto-L1"
        }
    }

    public static let stageOrder: [LightStaff] = [
        .sopranoL1,
        .sopranoL2,
        .tenorL,
        .bassL,
        .altoL2,
        .altoL1
    ]
}

// MARK: - Primer Assignment ---------------------------------------------------
public struct PrimerAssignment: Decodable {
    public let sample: String?
    public let note: String?

    /// Returns the filename (including subdirectory) that should be sent to clients.
    public var oscFileName: String? {
        guard let sample else { return nil }
        return sample.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    /// Returns the filename within the app bundle for local playback, normalising
    /// the short/long prefix to match the capitalised resource filenames.
    public var normalizedMacFileName: String? {
        guard var sample = oscFileName else { return nil }
        if sample.hasPrefix("primerTones/") {
            let suffix = sample.dropFirst("primerTones/".count)
            sample = String(suffix)
        }
        let lowered = sample.lowercased()
        if lowered.hasPrefix("short") {
            let numberPart = lowered.dropFirst("short".count)
            return "primerTones/Short\(numberPart)"
        } else if lowered.hasPrefix("long") {
            let numberPart = lowered.dropFirst("long".count)
            return "primerTones/Long\(numberPart)"
        }
        return "primerTones/\(sample)"
    }
}

// MARK: - Primer Color --------------------------------------------------------
/// Represents the nine Light Chorus colour groups used throughout the console.
public enum PrimerColor: String, CaseIterable, Codable, Identifiable {
    case blue
    case red
    case green
    case purple
    case yellow
    case pink
    case orange
    case magenta
    case cyan

    public var id: String { rawValue }

    /// Human-friendly display label.
    public var displayName: String {
        switch self {
        case .blue: return "Blue"
        case .red: return "Red"
        case .green: return "Green"
        case .purple: return "Purple"
        case .yellow: return "Yellow"
        case .pink: return "Pink"
        case .orange: return "Orange"
        case .magenta: return "Magenta"
        case .cyan: return "Cyan"
        }
    }

    /// Associated Light Chorus group index (1-based) used elsewhere in the app.
    public var groupIndex: Int {
        switch self {
        case .blue: return 1
        case .red: return 2
        case .green: return 3
        case .purple: return 4
        case .yellow: return 5
        case .pink: return 6
        case .orange: return 7
        case .magenta: return 8
        case .cyan: return 9
        }
    }

    /// Returns the stereo placement group for bundled playback on macOS.
    public var panPosition: PrimerPanPosition {
        switch self {
        case .blue, .red, .green: return .left
        case .purple, .yellow, .pink: return .center
        case .orange, .magenta, .cyan: return .right
        }
    }
}

// MARK: - Pan Position --------------------------------------------------------
public enum PrimerPanPosition {
    case left
    case center
    case right
}

File: flashlights_client/lib/model/event_recipe.dart
import 'dart:convert';
import 'package:flutter/services.dart' show rootBundle;

enum PrimerColor {
  blue,
  red,
  green,
  purple,
  yellow,
  pink,
  orange,
  magenta,
  cyan,
}

extension PrimerColorDisplay on PrimerColor {
  String get displayName {
    switch (this) {
      case PrimerColor.blue:
        return 'Blue';
      case PrimerColor.red:
        return 'Red';
      case PrimerColor.green:
        return 'Green';
      case PrimerColor.purple:
        return 'Purple';
      case PrimerColor.yellow:
        return 'Yellow';
      case PrimerColor.pink:
        return 'Pink';
      case PrimerColor.orange:
        return 'Orange';
      case PrimerColor.magenta:
        return 'Magenta';
      case PrimerColor.cyan:
        return 'Cyan';
    }
  }

  String get voicePart {
    switch (this) {
      case PrimerColor.green:
        return 'Sop-L1';
      case PrimerColor.magenta:
        return 'Sop-L2';
      case PrimerColor.orange:
        return 'Sop-L2';
      case PrimerColor.blue:
        return 'Alto-L1';
      case PrimerColor.red:
        return 'Alto-L2';
      case PrimerColor.cyan:
        return 'Alto-L2';
      case PrimerColor.yellow:
        return 'Ten-L';
      case PrimerColor.pink:
        return 'Bass-L';
      case PrimerColor.purple:
        return 'Bass-L';
    }
  }

  int get groupIndex {
    switch (this) {
      case PrimerColor.blue:
        return 1;
      case PrimerColor.red:
        return 2;
      case PrimerColor.green:
        return 3;
      case PrimerColor.purple:
        return 4;
      case PrimerColor.yellow:
        return 5;
      case PrimerColor.pink:
        return 6;
      case PrimerColor.orange:
        return 7;
      case PrimerColor.magenta:
        return 8;
      case PrimerColor.cyan:
        return 9;
    }
  }
}

enum ChoirFamily { soprano, alto, tenorBass }

ChoirFamily? choirFamilyFromString(String raw) {
  switch (raw.toLowerCase()) {
    case 'soprano':
      return ChoirFamily.soprano;
    case 'alto':
      return ChoirFamily.alto;
    case 'tenor_bass':
      return ChoirFamily.tenorBass;
    default:
      return null;
  }
}

extension ChoirFamilyAsset on ChoirFamily {
  String get recipeKey {
    switch (this) {
      case ChoirFamily.soprano:
        return 'soprano';
      case ChoirFamily.alto:
        return 'alto';
      case ChoirFamily.tenorBass:
        return 'tenor_bass';
    }
  }
}

PrimerColor? primerColorFromString(String raw) {
  switch (raw.toLowerCase()) {
    case 'blue':
      return PrimerColor.blue;
    case 'red':
      return PrimerColor.red;
    case 'green':
      return PrimerColor.green;
    case 'purple':
      return PrimerColor.purple;
    case 'yellow':
      return PrimerColor.yellow;
    case 'pink':
      return PrimerColor.pink;
    case 'orange':
      return PrimerColor.orange;
    case 'magenta':
      return PrimerColor.magenta;
    case 'cyan':
      return PrimerColor.cyan;
    default:
      return null;
  }
}

class PrimerAssignment {
  PrimerAssignment({required this.sample, this.note});

  final String sample;
  final String? note;

  String get normalizedSample => sample.trim();
}

class ElectronicsAssignment {
  ElectronicsAssignment({
    required this.sample,
    required this.channelMode,
    this.sourceStartMs,
    this.sourceEndMs,
    this.durationMs,
    this.fadeInMs,
    this.fadeOutMs,
    this.timingRule,
  });

  final String sample;
  final String channelMode;
  final double? sourceStartMs;
  final double? sourceEndMs;
  final double? durationMs;
  final double? fadeInMs;
  final double? fadeOutMs;
  final String? timingRule;
}

String? _canonicalPrimerSample(String? raw) {
  if (raw == null) {
    return null;
  }
  var value = raw.trim();
  if (value.isEmpty) {
    return null;
  }
  if (value.startsWith('./')) {
    value = value.substring(2);
  }
  const prefix = 'primerTones/';
  if (value.toLowerCase().startsWith(prefix.toLowerCase())) {
    value = value.substring(prefix.length);
  }
  var fileName = value;
  final lower = fileName.toLowerCase();
  if (lower.startsWith('short')) {
    fileName = 'Short${lower.substring(5)}';
  } else if (lower.startsWith('long')) {
    fileName = 'Long${lower.substring(4)}';
  }
  if (!fileName.toLowerCase().endsWith('.mp3')) {
    fileName = '$fileName.mp3';
  }
  return '$prefix$fileName';
}

String? _canonicalElectronicsSample(String? raw) {
  if (raw == null) {
    return null;
  }
  var value = raw.trim();
  if (value.isEmpty) {
    return null;
  }
  if (value.startsWith('./')) {
    value = value.substring(2);
  }
  if (!value.toLowerCase().endsWith('.mp3')) {
    value = '$value.mp3';
  }
  if (!value.startsWith('available-sounds/')) {
    value = 'available-sounds/$value';
  }
  return value;
}

class EventRecipe {
  EventRecipe({
    required this.id,
    required this.measure,
    required this.position,
    required this.primerAssignments,
    required this.electronicsAssignments,
  });

  final int id;
  // Official trigger-score measure for this event, not the sung-note onset.
  final int? measure;
  // Official trigger-score beat position for this event, not the sung-note onset.
  final String? position;
  final Map<PrimerColor, PrimerAssignment> primerAssignments;
  final Map<ChoirFamily, ElectronicsAssignment> electronicsAssignments;

  static EventRecipe fromJson(Map<String, dynamic> json) {
    final primer = <PrimerColor, PrimerAssignment>{};
    final primerJson = json['primer'] as Map<String, dynamic>?;
    if (primerJson != null) {
      primerJson.forEach((key, value) {
        final color = primerColorFromString(key);
        if (color != null && value is Map<String, dynamic>) {
          final sample = _canonicalPrimerSample(value['sample'] as String?);
          final note = (value['note'] as String?)?.trim();
          if (sample != null && sample.isNotEmpty) {
            primer[color] = PrimerAssignment(sample: sample, note: note);
          }
        }
      });
    }

    final electronics = <ChoirFamily, ElectronicsAssignment>{};
    final electronicsJson = json['electronics'] as Map<String, dynamic>?;
    if (electronicsJson != null) {
      electronicsJson.forEach((key, value) {
        final family = choirFamilyFromString(key);
        if (family == null || value is! Map<String, dynamic>) {
          return;
        }
        final sample = _canonicalElectronicsSample(value['sample'] as String?);
        final channelMode = (value['channelMode'] as String?)?.trim();
        if (sample == null ||
            sample.isEmpty ||
            channelMode == null ||
            channelMode.isEmpty) {
          return;
        }
        electronics[family] = ElectronicsAssignment(
          sample: sample,
          channelMode: channelMode,
          sourceStartMs: (value['sourceStartMs'] as num?)?.toDouble(),
          sourceEndMs: (value['sourceEndMs'] as num?)?.toDouble(),
          durationMs: (value['durationMs'] as num?)?.toDouble(),
          fadeInMs: (value['fadeInMs'] as num?)?.toDouble(),
          fadeOutMs: (value['fadeOutMs'] as num?)?.toDouble(),
          timingRule: (value['timingRule'] as String?)?.trim(),
        );
      });
    }

    return EventRecipe(
      id: json['id'] as int,
      measure: json['measure'] as int?,
      position: json['position'] as String?,
      primerAssignments: primer,
      electronicsAssignments: electronics,
    );
  }
}

class EventRecipeBundle {
  EventRecipeBundle({required this.events});

  final List<EventRecipe> events;

  static EventRecipeBundle fromJson(Map<String, dynamic> json) {
    final eventsJson = json['events'] as List<dynamic>? ?? const [];
    final events = eventsJson
        .whereType<Map<String, dynamic>>()
        .map(EventRecipe.fromJson)
        .toList(growable: false);
    return EventRecipeBundle(events: events);
  }
}

Future<List<EventRecipe>> loadEventRecipesAsset() async {
  final raw = await rootBundle.loadString('assets/event_recipes.json');
  final decoded = json.decode(raw) as Map<String, dynamic>;
  return EventRecipeBundle.fromJson(decoded).events;
}
